src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/azure/AzureManagerImpl.java

```diff
@@ -0,0 +1,2160 @@
+            }
+        } finally {
+            authDataLock.readLock().unlock();
+        }
+    }
+
+    private void updateSubscription(@NotNull Subscription subscription, @NotNull SSLSocketFactory sslSocketFactory)
+            throws AzureCmdException {
+        authDataLock.readLock().lock();
+
+        try {
+            String subscriptionId = subscription.getId();
+            ReentrantReadWriteLock subscriptionLock = getSubscriptionLock(subscriptionId, true);
+            subscriptionLock.writeLock().lock();
+
+            try {
+                if (subscriptions.containsKey(subscriptionId)) {
+                    Subscription existingSubscription = subscriptions.get(subscriptionId);
+                    existingSubscription.setManagementCertificate(subscription.getManagementCertificate());
+                    existingSubscription.setServiceManagementUrl(subscription.getServiceManagementUrl());
+                } else {
+                    subscriptions.put(subscriptionId, subscription);
+                }
+
+                setSSLSocketFactory(subscriptionId, sslSocketFactory);
+                storeSubscriptions();
+            } finally {
+                subscriptionLock.writeLock().unlock();
+            }
+        } finally {
+            authDataLock.readLock().unlock();
+        }
+    }
+
+    @Nullable
+    public UserInfo getUserInfo() {
+        authDataLock.readLock().lock();
+
+        try {
+            return userInfo;
+        } finally {
+            authDataLock.readLock().unlock();
+        }
+    }
+
+    private void setUserInfo(@Nullable UserInfo userInfo) {
+        authDataLock.writeLock().lock();
+
+        try {
+            this.userInfo = userInfo;
+            userInfoBySubscriptionId.clear();
+            removeUnusedSubscriptions();
+
+            storeSubscriptions();
+            storeUserInfo();
+        } finally {
+            authDataLock.writeLock().unlock();
+        }
+    }
+
+    private boolean hasSubscription(@NotNull String subscriptionId) {
+        authDataLock.readLock().lock();
+
+        try {
+            ReadWriteLockBox rwLockBox = new ReadWriteLockBox();
+
+            if (!getSubscriptionLock(subscriptionId, rwLockBox)) {
+                return false;
+            }
+
+            ReadWriteLock subscriptionLock = rwLockBox.getRWLock();
+            subscriptionLock.readLock().lock();
+
+            try {
+                return subscriptions.containsKey(subscriptionId);
+            } finally {
+                subscriptionLock.readLock().unlock();
+            }
+        } finally {
+            authDataLock.readLock().unlock();
+        }
+    }
+
+    @NotNull
+    private Subscription getSubscription(@NotNull String subscriptionId)
+            throws AzureCmdException {
+        authDataLock.readLock().lock();
+
+        try {
+            ReentrantReadWriteLock subscriptionLock = getSubscriptionLock(subscriptionId, false);
+            
```

Comment:
I'll rename getSSLSocketFactory to getSSLSocketFactoryIfPresent, returning Nullable, and check for null.
Whenever there is a 'has' method I prefer the associated 'get' method return NotNull, for internal consistency. In this case, I cannot remove the 'has' method, since it's being independently used many times.